### PR TITLE
web binding V2.0

### PR DESCRIPTION
**Pull request by mrrostam**
- updated the web binding
- updated the demo
- updated the unittest to use Selenium instead of Node
- updated both the main readme and the packages' readme

### CODE REVIEW

File: demo/web/index.html [hunk 1]
```diff
@@ -9,69 +9,80 @@
...
+            {
+              builtin: "Manzana",
+              sensitivity: 0.6,
+            },
+            porcupineKeywordCallback,
+            porcupineErrorCallback
+          );
+          let ppnFr = await PorcupineWebFrWorker.PorcupineWorkerFactory.create(
+            accessKey,
+            ["Framboise"],
```
Comment (ErisMik) [inline demo/web/index.html:67]: Why is this one different from the other languages?

**Review by ErisMik — APPROVED** [review]

File: demo/web/index.html (hunk 1 above)
Comment (mrrostam) [reply demo/web/index.html:67]: there are different ways to create the worker and pass keywords to the engine. I think the idea here was to show different possibilities (I got this part from the old demo written by David)

Comment (ErisMik) [reply demo/web/index.html:67]: Okay, I figured, just wanted to make sure it wasn't a mistake.

File: README.md [hunk 2]
```diff
@@ -1166,25 +1166,21 @@ Each spoken language is available as a dedicated npm package (e.g. @picovoice/po
     <script src="https://unpkg.com/@picovoice/porcupine-web-en-worker/dist/iife/index.js"></script>
     <script src="https://unpkg.com/@picovoice/web-voice-processor/dist/iife/index.js"></script>
     <script type="application/javascript">
+      function keywordDetectionCallback(keyword) {
+        console.log("Porcupine detected " + keyword);
```
Comment (kenarsa) [inline README.md:1170]: I wonder if this the cool JS way of doing it?

```javascript
`Porcupine detected  {keyword}`
```

I actually am not sure about all implications

File: binding/web/scripts/copy_wasm.js [hunk 3]
```diff
@@ -22,13 +22,15 @@ for (const language of LANGUAGES) {
...
   const outputDirectory = join(__dirname, "..", language);
 
-  ncp(sourceDirectory, outputDirectory, (error) => {
-    if (error) {
-      console.error(error);
-    } else {
-      console.log(`Copied ${sourceDirectory} => ${outputDirectory}`);
-    }
-  });
+  fs.mkdirSync(outputDirectory, { recursive: true });
```
Comment (kenarsa) [inline binding/web/scripts/copy_wasm.js:25]: we prob want something to catch errors? Otherwise, it would be silent?

File: README.md [hunk 4]
```diff
@@ -1166,25 +1166,21 @@ Each spoken language is available as a dedicated npm package (e.g. @picovoice/po
     <script src="https://unpkg.com/@picovoice/porcupine-web-en-worker/dist/iife/index.js"></script>
     <script src="https://unpkg.com/@picovoice/web-voice-processor/dist/iife/index.js"></script>
     <script type="application/javascript">
+      function keywordDetectionCallback(keyword) {
```
Comment (kenarsa) [inline README.md:1169]: might be a good idea to include errorCallback to these snippets as well.

File: README.md (hunk 2 above)
Comment (mrrostam) [reply README.md:1170]: you are absolutely right

File: .github/workflows/web.yml [hunk 5]
```diff
@@ -22,20 +22,34 @@ defaults:
...
 jobs:
   build:
-    runs-on: ubuntu-latest
+    runs-on: ubuntu-20.04
 
     strategy:
       matrix:
-        node-version: [14.x, 16.x]
+        python-version: [3.8]
+        node-version: [12.x]
```
Comment (kenarsa) [inline .github/workflows/web.yml]: let's use LTS

File: README.md [hunk 6]
```diff
@@ -1224,22 +1225,23 @@ npm install @picovoice/porcupine-web-en-worker @picovoice/web-voice-processor
 import { PorcupineWorkerFactory } from "@picovoice/porcupine-web-en-worker"
 import { WebVoiceProcessor } from "@picovoice/web-voice-processor"
 
+function keywordDetectionCallback(keyword) {
+  console.log("Porcupine detected " + keyword);
```
Comment (kenarsa) [inline README.md:1229]: interpolation please

File: binding/web/template/README.md [hunk 7]
```diff
@@ -60,27 +69,30 @@ yarn add @picovoice/web-voice-processor @picovoice/porcupine-web-en-worker
...
 
-async startPorcupine()
+// The worker will call the callback function upon a detection event
+// Here we tell it to log it to the console
+function keywordDetectionCallback(keyword) {
+  console.log(`Porcupine detected ${keyword}`);
+}
+
+// If during processing an audio stream, anything goes wrong, the worker will
+// call the error callback and passe the error to it.
```
Comment (kenarsa) [inline binding/web/template/README.md:79]: passe >> pass
can you please run spellcheck on this PR? hard to find manually.

File: binding/web/template/src/porcupine.ts [hunk 8]
```diff
@@ -130,18 +171,21 @@ class Porcupine implements PorcupineEngine {
    * Behind the scenes, it requires the WebAssembly code to load and initialize before
    * it can create an instance.
    *
+   * @param accessKey - AccessKey obtained from Picovoice Console
```
Comment (kenarsa) [inline binding/web/template/src/porcupine.ts:174]: link please

**Review by kenarsa — CHANGES_REQUESTED** [review]

**Review by ErisMik — APPROVED** [review]

File: binding/web/template/src/porcupine_worker.ts [hunk 9]
```diff
@@ -138,7 +109,23 @@ onmessage = function (
       Porcupine.rejectFilePromise(event.data.message);
       Porcupine.clearFilePromises();
       break;
-    default:
-      internalOnmessage(event.data);
+      case 'init':
+        init(event.data.accessKey, event.data.keywords,event.data.start);
```
Comment (ErisMik) [inline binding/web/template/src/porcupine_worker.ts:113]: Are we sure we don't want to wrap the `init()` in a mutex also?

**Review by ErisMik — APPROVED** [review]

**Pull request merged**